_posts/2018-05-24-Spring-IO-2018.md

```diff
@@ -0,0 +1,777 @@
+
+Transactional semantics are still retained because the eventing is synchronous, by default.
+This also applies for JEE eventing.
+
+The [@TransactionalEventListener](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html){:target="_blank"} annotation allows you to delay the execution of events, so for example, you can send out an email when an Order has truly been completed.
+
+> Side Step: Error Scenarios
+
+When a synchronous event listener fails, this will be handled by the transaction, so no worries.
+
+But when an asynchronous event listener fails, the transaction does not get rolled back and you will need to deal with retries.
+
+You can make use of an Event Publication Registry when you use [TransactionalEventListeners](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html){:target="_blank"} as these event listeners are decorated with a log before the commit, since the system needs to know the events need to be sent out to.
+When the event has been processed, the log will be cleared.
+If it doesn't get cleared the system can keep retrying, so you don't lose events.
+
+Example code of the modulith can be found [here](https://github.com/olivergierke/sos/tree/master/20-modulith){:target="_blank"}
+
+### System of Systems
+
+#### Messaging
+
+Whenever you make use of a message broker, you introduce a potentional single point of failure, like with [Apache Kafka](https://kafka.apache.org/){:target="_blank"} or [RabbitMQ](https://www.rabbitmq.com/){:target="_blank"}.
+These brokers know about all the messages of all the systems and decide how long these messages will be retained.
+
+Coupling does exist, although not explicit, but the message format will decide which version of a service can process these messages, just as with REST.
+Especially if you keep your events for a long time, which is possible with Kafka, you might need to think about transforming existing events.
+
+But these messaging systems tend to be designed for scale.
+
+Pro tip: make use of [JsonPath](https://github.com/json-path/JsonPath){:target="_blank"} annotations for the message payload in order to make it more robust.
+
+Example code of a system of systems with messaging can be found [here](https://github.com/olivergierke/sos/tree/master/30-messaging-sos){:target="_blank"}
+
+#### REST Messaging
+
+If you use REST you will have to deal with caching, pagination and conditional requests.
+
+Messages do not tend to be stored for long periods of time and most communication tends to be synchronous.
+One does have to pay attention on not to lose events as your application will immediately know if the message was processed correctly, incorrectly or timed out.
```

Comment:
"One does have to pay attention on not to lose events" doesn't sound right. How about 
"One does have to pay attention on not losing events" or turn it positive: "One does have to pay attention on retaining events"?